Add v3 onion support to tor-hidden-services ansible role

kushaldas opened the pull request:
## Status

Ready for review

## Description of Changes

Fixes #4628 
Fixes #4667

Started adding related Ansible changes to have v3 onion services (including the authenticated ones).

## Testing

### v3 only

- [ ] Fresh install
- [ ] Rerun `./securedrop-admin install` on a already installed system.

For both the options above, make the following two changes in `install_files/ansible-base/group_vars/all/site-specific`

```
v2_onion_services: false
v3_onion_services: true
```

For this case:

- [ ] No v2 address working and or in the system
- [ ] v3 source address works
- [ ] v3 journalist interface works (follow <https://kushaldas.in/posts/setting-up-authorized-v3-onion-services.html> about setting up Torrc)

### v2/v3 alongside

```
v2_onion_services: true
v3_onion_services: true
```

For this:

- [ ] All old v2 address working and are in the `app` and `mon` servers 
- [ ] v3 source address works
- [ ] v3 journalist interface works (follow <https://kushaldas.in/posts/setting-up-authorized-v3-onion-services.html> about setting up Torrc)

### v2 only

There is also the third point to test which is not adding `v2_onion_services` or `v3_onion_services` to `site-specific` and instead the default values (v2 on, v3 off).

This should keep the current v2 systems as it is, and no v3 details should be generated in the servers.

## Deployment

Any special considerations for deployment? Consider both:

1. Upgrading existing production instances.
2. New installs.

## Checklist

### If you made changes to the server application code:

- [ ] Linting (`make lint`) and tests (`make -C securedrop test`) pass in the development container

### If you made changes to `securedrop-admin`:

- [ ] Linting and tests (`make -C admin test`) pass in the admin development container

### If you made changes to the system configuration:

- [ ] [Configuration tests](https://docs.securedrop.org/en/latest/development/testing_configuration_tests.html) pass

### If you made non-trivial code changes:

- [ ] I have written a test plan and validated it for this PR

### If you made changes to documentation:

- [ ] Doc linting (`make docs-lint`) passed locally

eloquence commented:
(Since it's a draft PR and cannot be merged anyway, I've removed "DO NOT MERGE" from the PR title)

kushaldas commented:
While working on `tailsconfig` I figured one issue with the current logic.

In v3, we have to create at least 2 different set of `public/private` keys, one for the journalist UI, and the other for `ssh` service if required.

@redshiftzero Is it okay if I use same set of keys for both `app` and `mon` servers, or should we create 2 different sets for both the servers?

redshiftzero commented:
@kushaldas I think we should use separate keys where possible, so separate keys for the SSH services on app and mon

conorsch commented on admin/requirements.in:
```diff
@@ -1,2 +1,3 @@
 prompt_toolkit
 pyyaml
+cryptography==2.7
```
Can you squash the commits related to this version bump? It'd be great have the code diff for the key generation (which requires the version bump) in the same commit as the version bump. It's also not clear which parts of cryptography needed to be updated—presumably the x25519 calls? Documenting that in the commit message will be hugely helpful to future maintainers.

kushaldas replied:
Yes, I will do that.

conorsch replied:
When resolving the reqs commits, consider also appending the cryptography bump to the dev env vars: if the script is executed by Ansible (see rationale in https://github.com/freedomofpress/securedrop/pull/4652#discussion_r312283362), then it must have a recent version of cryptography available (or else we use bash). More detail on _why_ the version bump to crytpography is required here would be helpful in determining the best way to generate the keypairs!

conorsch commented on admin/securedrop_admin/__init__.py:
```diff
@@ -544,10 +548,68 @@ def sdconfig(args):
+    public/private key pair.
+    """
+    secret_key_path = os.path.join(args.ansible_path,
+                                   "tor_v3_keys.json")
```
Just as we currently have two discrete files on the Admin Workstation for the v2 auth info, it'd be great to have the same for the v3 info. Doing so would simplify the lookup logic required in the dynamic-inventory script (not yet implemented in this PR), as well. Perhaps the function should receive `hostname=None`, and select the files that way.

kushaldas replied:
I feel that one file for admin workstation with all 3 different key pairs (journalist, ssh app, and ssh mon) and a file for journalist workstations (only with the journalist keypair) is simpler. Also, this is only for the key information, if we want, we can also remove the `ths` files all together and just the `.onion` address from the JSON files iteself (we have to add them there).

conorsch replied:
Having taken a closer look at the keygen logic, we're still going to need separate files for the dynamic-inventory support (OK to defer to #4629). With the v2 logic, if-and-only-if the `-ths` files are present on the Admin Workstation, then Onion URLs are used for the SSH connection. Since those `-ths` files are created on the Admin Workstation at the _end_ of the install process, the logic is sound. For the v3 keygen logic, however, we need a way to denote whether the corresponding Onion URLs should be used for the SSH connection. 

I suggest moving the keygen logic into Ansible via a script, so that the JSON file does not exist when Ansible builds the SSH connection, meaning local IPv4 addresses (or v2 Onion URLs) will be used. At the end of the run, we can sprinkle in appropriate `-v3-ths` files back onto the Admin Workstation.

We cannot relegate the v3 keygen logic solely to the `securedrop-admin` init, because that's only relevant for Tails environments, which means developer staging VMs will _not_ have v3 connectivity.

conorsch commented on install_files/ansible-base/group_vars/securedrop_application_server.yml:
```diff
@@ -20,6 +20,17 @@ tor_instances:
     filename: app-journalist-aths
 
+tor_instances_v3:
+  - "{{ {'service': 'sshv3', 'filename': 'app-sshv3-aths'} if enable_ssh_over_tor else [] }}"
```
The `else []` causes an empty list here—is that intentional? The other elements of `tor_instances_v3` are dicts, not lists. Haven't tested functionally yet to observe an error here, but it seems inconsistent.

kushaldas replied:
I copied that from the above lines in the same file https://github.com/freedomofpress/securedrop/blame/07fed0c56e89d7edb006a7aa9e0f850d3d516216/install_files/ansible-base/group_vars/securedrop_application_server.yml#L17  Seems to be added by @msheiny  2 years ago.

conorsch commented on install_files/ansible-base/roles/restrict-direct-access/tasks/fetch_tor_config.yml:
```diff
@@ -5,6 +5,7 @@
     path: "{{ tor_hidden_services_parent_dir }}/{{ item.service }}/hostname"
     delay: 5
   with_items: "{{ tor_instances }}"
+  when: "v2_onion_services|default(True)"
```
Rather than repeating `|default(True)`, set `v2_onion_services: True` in `install_files/ansible-base/group_vars/all/securedrop`.

kushaldas replied:
I will do this.

conorsch commented on install_files/ansible-base/roles/restrict-direct-access/tasks/fetch_tor_config.yml:
```diff
@@ -26,6 +28,42 @@
+  # Read-only task, so don't report changed.
+  changed_when: false
+  with_items: "{{ tor_instances_v3 }}"
+  when: "v3_onion_services|default(False)"
```
As above, rather than repeating `|default(False)`, set `v3_onion_services: False` in `install_files/ansible-base/group_vars/all/securedrop`. Then the `sdconfig` command can update that value to True, in line with the recommendations in #4628.

kushaldas replied:
I will do this.

conorsch commented on admin/securedrop_admin/__init__.py:
```diff
@@ -544,10 +548,68 @@ def sdconfig(args):
+    with open(filepath) as fobj:
+        data = json.load(fobj)
+
+    return data
```
Nit: don't use `data` as a variable name—all variables contain data! :slightly_smiling_face:

kushaldas replied:
I will update.

redshiftzero replied:
resolved in `v3 onions: remove magic number for removing base32 padding characters`

conorsch commented on install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml:
```diff
@@ -43,3 +55,137 @@
+    - "v3_onion_services|default(False)"
+    - inventory_hostname == 'app'
+  tags:
+    - tor
```
The key-management tasks here are largely duplicated, with conditional checks for determining whether app or mon is the current host. Since the tasks will by default run against _both_ hosts, and the facts will be set uniquely per-host, we don't need the duplication.

conorsch commented on install_files/ansible-base/roles/tor-hidden-services/templates/client_auth_app_ssh.j2:
```diff
@@ -0,0 +1 @@
+descriptor:x25519:{{ tor_v3_app["app_ssh_key"] }}
```
As above, these templates don't need to be unique per-host—see for reference the existing v2 logic:

https://github.com/freedomofpress/securedrop/blob/20cbd52cb40ce1cef09407f81fc338bf6765aaf9/install_files/ansible-base/roles/restrict-direct-access/tasks/fetch_tor_config.yml#L21-L31

conorsch commented:
Appended two commits, aiming to simplify the keypair-management logic within the Ansible roles. Please take a look, @kushaldas, and see if you agree that the terser version is suitable.

While working on this, I noticed that the branch here is outdated; it doesn't include the related Tor logic introduced in #4648:

```
$ git checkout develop
Switched to branch 'develop'
Your branch is up to date with 'origin/develop'.
$ grep -rF update_onion_version_config
admin/securedrop_admin/__init__.py:        self.update_onion_version_config()
admin/securedrop_admin/__init__.py:    def update_onion_version_config(self):
admin/tests/test_securedrop-admin.py:        site_config.update_onion_version_config()
admin/tests/test_securedrop-admin.py:        site_config.update_onion_version_config()
admin/tests/test_securedrop-admin.py:        site_config.update_onion_version_config()
[user@sd-dev:~/securedrop] [sd] develop+* ±
$ git checkout onion_onion_and_more_onions
Switched to branch 'onion_onion_and_more_onions'
Your branch is up to date with 'origin/onion_onion_and_more_onions'.
$ grep -rF update_onion_version_config
$ # nothing found
```

If you agree the added commits are work keeping, please rebase on top of latest develop, so we're testing with the most recently merged work.

The most important outstanding changes in my view are:

  * justify version bump to "cryptography" dependency
  * move v3 keygen logic out of `securedrop-admin`, otherwise v3 won't work for devs

Based on the discussion in https://github.com/freedomofpress/securedrop/issues/4628#issuecomment-516494821, it looks like you've already tried to the script-based approach. Let's try to get that working again. See example below for a slightly modified version of your script. If we can adjust the calls to get it working with the currently used version of cryptography, great, otherwise let's make sure to bump the requirement in the same contexts Ansible is used (i.e. also dev envs).


<details>

<summary>
Example script generate-tor-v3-keypairs
</summary>

```
diff --git a/install_files/ansible-base/roles/tor-hidden-services/files/generate-tor-v3-keypairs b/install_files/ansible-base/roles/tor-hidden-services/files/generate-tor-v3-keypairs
new file mode 100644
index 000000000..666eed601
--- /dev/null
+++ b/install_files/ansible-base/roles/tor-hidden-services/files/generate-tor-v3-keypairs
@@ -0,0 +1,51 @@
+#!/usr/bin/env python3
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric import x25519
+
+
+def generate_x25519_keypair():
+    """This function generate new keys and returns them as tuple.
+
+    :returns: Tuple(public_key, private_key)
+    """
+
+    private_key = x25519.X25519PrivateKey.generate()
+    private_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PrivateFormat.Raw,
+        encryption_algorithm=serialization.NoEncryption())
+    public_key = private_key.public_key()
+    public_bytes = public_key.public_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PublicFormat.Raw)
+
+    public = base64.b32encode(public_bytes)[:-4].decode("utf-8")
+    private = base64.b32encode(private_bytes)[:-4].decode("utf-8")
+    return public, private
+
+
+def generate_new_tor_v3_keypairs():
+    """
+    This method will either read the old keys or generate a new
+    public/private key pair.
+    """
+    # No old keys, generate and store them first
+    app_journalist_public_key, app_journalist_private_key = generate_x25519_keypair()
+    # For app ssh service
+    app_ssh_public_key, app_ssh_private_key = generate_x25519_keypair()
+    # For mon ssh service
+    mon_ssh__public_key, mon_ssh_private_key = generate_x25519_keypair()
+    tor_v3_service_info = {
+            "app_journalist_public_key": app_journalist_public_key,
+            "app_journalist_private_key": app_journalist_private_key,
+            "app_ssh_public_key": app_ssh_public_key,
+            "app_ssh_private_key": app_ssh_private_key,
+            "mon_ssh_public_key": mon_public_ssh_key,
+            "mon_ssh_private_key": mon_ssh_private_key,
+    }
+    # Send results to stdout
+    print(json.dumps(tor_v3_service_info))
+
+
+if __name__ == "__main__":
+    generate_new_tor_v3_keypairs()
```

</details>

Once we have passable keypair-and-service generation in this PR, we can discuss specifics on the Tails-specific integration in #4629. Let me know your thoughts on the above, @kushaldas.

kushaldas commented:
@conorsch I am okay with the approach you mentioned here. What should be the next step?

conorsch commented:
> Appended two commits

Oops, my mistake: I hadn't actually pushed! The two new commits are now present on the branch, see https://github.com/freedomofpress/securedrop/pull/4652/commits/fea52e0fe7739a7598640ee0248447c947af1a85 & https://github.com/freedomofpress/securedrop/pull/4652/commits/92e0949b66609706334ecad85506bdccc3c6cd85.

> What should be the next step?

@kushaldas Recapping my comment above, the next steps should be:

* review the two commits I added
* rebase this branch on develop, to include changes from #4648 and others
* explain why a version bump to the "cryptography" dependency is necessary; it's hard to understand in the correct branch _why_ this change was made
* move v3 keygen logic out of `securedrop-admin`, into a script runnable by Ansible (see example above), otherwise v3 URLs won't work for devs

Please handle those tasks, and ping me when you feel the branch is ready for another look!

kushaldas commented:
>move v3 keygen logic out of securedrop-admin, into a script runnable by Ansible (see example above), otherwise v3 URLs won't work for devs

@conorsch Last time when I tried, I could not do that the above in a nice `Ansible` way. I will need your help in that too.

msheiny commented:
>   move v3 keygen logic out of securedrop-admin, into a script runnable by Ansible (see example above), otherwise v3 URLs won't work for devs

What about tweaking the dev initialization entry-points to call a subset of securedrop admin scripts first? I don't think ansible is the right answer to every scriptable problem. Ansible is really good at running a bunch of tasks remotely. This task is run locally and would benefit from more in-depth unit testing. Recall that in the past at one point, the admin scripting was in ansible but that was deemed unmaintainable and was moved to python.

conorsch commented:
The currently implemented securedrop-admin strategy cannot configure v3 URLs when testing against VMs, so the invocation must change. It needn't be pure Ansible—simply calling a script makes sense—but at no point other than Tails-based QA testing are devs regularly invoking the securedrop-admin logic. The current script logic uses python `cryptography`, which means required libraries will be present within the virtualenv that executes both securedrop-admin and Ansible, on the Admin Workstation.

If we moved to a leaner method like bash, we might wind up with more portable keygen logic, but that doesn't seem worthwhile—we'd have to be careful about system deps like openssl versions. As far as _what_ calls the keygen logic script, the most important attributes are 1) it's automatic, so devs don't forget; 2) the same process is used on dev/admin workstations, as far as possible.

kushaldas commented:
>The currently implemented securedrop-admin strategy cannot configure v3 URLs when testing against VMs, so the invocation must change. 

Current means? The branch I pushed last week, did it amazingly :)

kushaldas commented:
Also, we can just make the dynamic inventory to call the similar python function to get key pairs while doing from `molecule`.

kushaldas commented on admin/requirements-ansible.in:
```diff
@@ -1,3 +1,3 @@
 ansible>2.6<2.7
-cryptography>=2.3
+cryptography>=2.5
```
New comment here? @conorsch

conorsch replied:
Ultimately, the explanation of _why_ we're bumping the version of cryptography should be in a commit message, but for now, go ahead and add a comment to the `.in` file—we can always pull that into a commit message later, as part of review!

redshiftzero replied:
resolved in `v3 onions: use cryptography>=2.7`

kushaldas commented:
@conorsch Added the changes requested. When I tried the scrip way, I could not find a way to get the correct `cryptography` into this script's environment.

conorsch commented:
>>The currently implemented securedrop-admin strategy cannot configure v3 URLs when testing against VMs, so the invocation must change.

>Current means? The branch I pushed last week, did it amazingly :)

I could have been more clear: the current strategy (accurate as of today on this branch) fails to configure v3 URLs _on staging VMs unless a Tails VM is also used_. That means that `make staging` will not configure v3 URLs, nor will the v3 service config be tested in CI. We'll need config tests for the v3 tor changes; consider writing those tests sooner rather than later to guide development!

> When I tried the scrip way, I could not find a way to get the correct cryptography into this script's environment.

That sounds like a consequence of the crytography version bump. Although you added the new cryptography version to `admin/requirements-ansible.in`, you must also run `make update-python{2,3}-requirements` for that dependency to propagate to the relevant `requirements.txt` files. Once you do that, the script should work just fine for you, modulo linting errors from the copy/paste version currently in place. I already pushed a small commit fixing yamllint errors on the new Ansible logic.

Testing locally, I was able to invoke the keygen script successfully via Ansible using the following command:

```
ansible -c local -i localhost, all -m script -a 'install_files/ansible-base/roles/tor-hidden-services/files/generate-tor-v3-keypairs'
```

After cleaning up the aforementioned linting errors, that script executes well, and returns JSON (albeit escaped in the interactive terminal, since Ansible renders task results as JSON by default). So, for next steps, please do the following:

1. Rebase this branch on top of latest develop, to include changes from #4648 and others.
2. Expand comment justifying cryptography version bump. (Remember, this is for future maintainers: which calls were not present in the old API? Why are those calls necessary?)
3. Update `*.txt` requirements files via the Makefile targets, then update your local venv to use the new reqs.
4. Modify the `generate-tor-v3-keypairs` script to execute successfully (has a few missing imports right now).
5. Run the script inside the Ansible tasks, if-and-only-if the JSON file doesn't exist locally. That will ensure the JSON info is present, and we can read it in for the host config, as the branch already does.
6. Write testinfra tests to validate to the tor v3 service state. These may need to wait until v3 is "on-by-default"; still, we should define "successful" service behavior in the tests, even if we temporarily disable them.

Based on previous discussion, there may be outstanding questions regarding the implementation of step 5. Please set aside some time to give it your best shot. If you get stuck and it makes sense for us to pair on Ansible work for a bit, we can find a time to do so.

kushaldas commented:
>1. Rebase this branch on top of latest develop, to include changes from #4648 and others.

Done.

>2. Expand comment justifying cryptography version bump. (Remember, this is for future maintainers: which calls were not present in the old API? Why are those calls necessary?)

I added the URL to the docs and updated the comment.

> 3. Update `*.txt` requirements files via the Makefile targets, then update your local venv to use the new reqs.

This was done before itself (part of the PR already).

> 4. Modify the `generate-tor-v3-keypairs` script to execute successfully (has a few missing imports right now).

This is the error I was talking about, that it will not work from Ansible as a command or script as it is missing the virtualenv.

@conorsch ^^

kushaldas commented on admin/requirements.txt:
```diff
@@ -70,36 +70,29 @@ cffi==1.11.4 \
+    --hash=sha256:cfee9164954c186b191b91d4193989ca994703b2fff406f71cf454a2d3c7327e \
+    --hash=sha256:e6347742ac8f35ded4a46ff835c60e68c22a536a8ae5c4422966d06946b6d4c6 \
+    --hash=sha256:f27d93f0139a3c056172ebb5d4f9056e770fdf0206c2f422ff2ebbad142e09ed \
+    --hash=sha256:f57b76e46a58b63d1c6375017f4564a28f19a5ca912691fd2e4261b3414b618d
```
@redshiftzero That would be nice.

redshiftzero replied:
roger that, will do today

redshiftzero replied:
pyca/cryptography 2.3 -> 2.7 diff review done: [wiki link](https://github.com/freedomofpress/securedrop-debian-packaging/wiki/cryptography-2.3-to-2.7) (full diff intentionally not posted as I have the source tarball hashes before and after signed), also posted to `diff-review@python.org`

kushaldas commented:
@conorsch  I am getting this:

```
 ___________________________________________________________
/ TASK [tor-hidden-services : Create directories for Tor v3 \
\ authenticated onion services.]                            /
 -----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

fatal: [app]: FAILED! => {"msg": "The conditional check ''source' not in item.service' failed. The error was: error while evaluating conditional ('source' not in item.service): Unable to look up a name or access an attribute in template string ({% if 'source' not in item.service %} True {% else %} False {% endif %}).\nMake sure your variable name does not contain invalid characters like '-': argument of type 'StrictUndefined' is not iterable\n\nThe error appears to have been in '/home/kdas/code/securedrop/install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml': line 25, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n\n- name: Create directories for Tor v3 authenticated onion services.\n  ^ here\n"}

```

kushaldas commented:
I am using the following patch to test the change I made. I have commented out validation part as I am testing this from my `debian` box to prod vms (instead of tails).


```patch
diff --git a/admin/securedrop_admin/__init__.py b/admin/securedrop_admin/__init__.py
index c71461790..bf5dad7cd 100755
--- a/admin/securedrop_admin/__init__.py
+++ b/admin/securedrop_admin/__init__.py
@@ -916,6 +916,10 @@ def parse_argv(argv):
                                             help=reset_admin_access.__doc__)
     parse_reset_ssh.set_defaults(func=reset_admin_access)
 
+    parse_tailsconfig = subparsers.add_parser('generate_v3_keys',
+                                              help=find_or_generate_new_torv3_keys.__doc__)
+    parse_tailsconfig.set_defaults(func=find_or_generate_new_torv3_keys)
+
     return set_default_paths(parser.parse_args(argv))
 
 
diff --git a/install_files/ansible-base/inventory-dynamic b/install_files/ansible-base/inventory-dynamic
index 0f2adbefe..97df5f871 100755
--- a/install_files/ansible-base/inventory-dynamic
+++ b/install_files/ansible-base/inventory-dynamic
@@ -13,7 +13,7 @@ import yaml
 # Constant for storing SSH connection. This is reused for all hosts,
 # but it's trivial to store the info in the primary dict to support
 # per-host vars.
-SSH_COMMON_ARGS = "-o ProxyCommand='nc -x 127.0.0.1:9050 %h %p'"
+SSH_COMMON_ARGS = ""
 
 # Find absolute path to ansible-base directory.
 SECUREDROP_ANSIBLE_DIRECTORY = os.path.abspath(os.path.join(__file__,
diff --git a/install_files/ansible-base/roles/tor-hidden-services/defaults/main.yml b/install_files/ansible-base/roles/tor-hidden-services/defaults/main.yml
index 6e86c95dd..132b257e5 100644
--- a/install_files/ansible-base/roles/tor-hidden-services/defaults/main.yml
+++ b/install_files/ansible-base/roles/tor-hidden-services/defaults/main.yml
@@ -2,3 +2,4 @@
 tor_hidden_services_parent_dir: /var/lib/tor/services
 tor_user: debian-tor
 enable_ssh_over_tor: true
+sd_root_dir: "{{ lookup('pipe','git rev-parse --show-toplevel') }}"
diff --git a/install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml b/install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml
index 966083487..f00798c63 100644
--- a/install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml
+++ b/install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml
@@ -49,6 +49,16 @@
   tags:
     - tor
 
+- name: Generate Onion v3 keys if required the Tails admin system
+  command: "{{ sd_root_dir }}/securedrop-admin generate-v3-keys"
+  delegate_to: localhost
+  # Local action, so we don't want elevated privileges
+  become: no
+  when: "v3_onion_services|default(False)"
+  tags:
+    - tor
+    - admin
+
 - name: Get the v3 keys locally from the Tails admin system
   set_fact:
     v3_local_key_info: "{{ lookup('file', role_path+'/../../tor_v3_keys.json')|from_json }}"
diff --git a/install_files/ansible-base/securedrop-prod.yml b/install_files/ansible-base/securedrop-prod.yml
index 6526173f9..1f6b669ce 100755
--- a/install_files/ansible-base/securedrop-prod.yml
+++ b/install_files/ansible-base/securedrop-prod.yml
@@ -1,14 +1,14 @@
 #!/usr/bin/env ansible-playbook
 ---
-- name: Ensure validation is run before prod install
-  hosts: localhost
-  environment:
-    LC_ALL: C
-  max_fail_percentage: 0
-  any_errors_fatal: yes
-  connection: local
-  roles:
-    - { role: validate, tags: validate }
+#- name: Ensure validation is run before prod install
+#  hosts: localhost
+#  environment:
+#    LC_ALL: C
+#  max_fail_percentage: 0
+#  any_errors_fatal: yes
+#  connection: local
+#  roles:
+#    - { role: validate, tags: validate }
 
 - name: Prepare servers for installation
   hosts: securedrop

```

redshiftzero commented on admin/securedrop_admin/__init__.py:
```diff
@@ -566,10 +570,71 @@ def sdconfig(args):
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PublicFormat.Raw)
+
+    public = base64.b32encode(public_bytes)[:-4].decode("utf-8")
```
resolved in commit `v3 onions: remove magic number for removing base32 padding characters` (using commit messages here for identifiers in case of likely future rebasing)

redshiftzero commented on install_files/ansible-base/roles/tor-hidden-services/tasks/configure_tor_hidden_services.yml:
```diff
@@ -29,7 +29,7 @@
     group: "{{ tor_user }}"
     mode: "0700"
-  with_items: "{{ tor_auth_instances_v3 }}"
+  with_items: "{{ tor_instances_v3 }}"
```
@kushaldas this change resolves the problem you describe here: https://github.com/freedomofpress/securedrop/pull/4652#issuecomment-521395752

redshiftzero commented:
I took a spin through this this evening and have v2/v3 working alongside each other locally on the changes in your PR in the staging environment (for this testing I'm _not_ using Tails in order to verify staging continues to work for devs, see note in next paragraph) :tada: 

One note/TODO: we do need to install the admin requirements in the staging env for the task in 077fcbf27b132fcfb310a9a513f415bdb613843b - but since the installation of the admin requirements is a dev/CI change only, I don't see an issue with doing this, let me know if you see a problem there (for prod the admin env is already the active virtualenv so no change is required there).

Otherwise the changes we need are test coverage for the additions to `securedrop-admin` as well as `testinfra` tests (since part of the changes I appended include enabling v3 in staging, so we can now add configuration tests to assert that the tasks are doing what we expect) - want to take a spin at least the former @kushaldas?

kushaldas commented:
When I tried to re-run install:

```
$ ./securedrop-admin install
Tor v3 onion service keys already exist in: /home/kdas/code/securedrop/install_files/ansible-base/tor_v3_keys.json

```

zenmonkeykstop commented on install_files/ansible-base/roles/restrict-direct-access/templates/ths_config_v3.j2:
```diff
@@ -0,0 +1 @@
+{% if item.item.filename.endswith('-aths') %}HidServAuth {% endif %}{{ item.stdout }}
```
AFAICT the HidServAuth directive isn't used for v3 auth services. It might be better to create the client-side auth files here instead. So the source interface `app-source-v3ths` file would look like 
```
<onionUrl>
```
and the three `app-journalist-v3aths, app-ssh-v3aths, mon-ssh-v3aths` file would look like
```
<onionURLFirst56>:<(journalistv3|sshv3|monv3)>:x25519:<privateKey> 
```
This would a) simplify the `tailsconfig` logic and b) make the staging workflow for getting at the authenticated services roughly similar to the existing one for v2. Right now devs need to append the aths contents to a torrc (browser or daemon), for v3 they would have to copy the v3aths files to their ClientOnionAuthDir.

kushaldas replied:
The journalist keys will then have to copy more files, that is the reason I wanted to separate it out totally.

zenmonkeykstop replied:
Nope, I think all the journalist workstation would need would be the `app-journalist-v3aths` file, which `tailsconfig` would copy into place and parse for the onion URL for desktop shortcuts, which is basically what's happening already (except the `HidServAuth` line is appended to `torrc` rather than being copied as a file somewhere).

zenmonkeykstop replied:
Setting up a new Admin key _would_ require that the JSON file be copied over as well, along with `site-specific` and all the `*ths` files, but I think that we should be recommending to people that they make new Admin keys by cloning the PV from a backup and rotating the new key's SSH creds anyway.

kushaldas commented:
I did a fresh v2 install from the branch. And then enabled v3 in the `site-specific`, and then re-run `./securedrop-admin install` command. It is now stuck at following...

```
 __________________________________________________________
/ TASK [restrict-direct-access : Wait for all Tor v3 onion \
\ services hostname files.]                                /
 ----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

failed: [app] (item={u'service': u'sourcev3', u'filename': u'app-sourcev3-ths'}) => {"changed": false, "elapsed": 300, "item": {"filename": "app-sourcev3-ths", "service": "sourcev3"}, "msg": "Timeout when waiting for file /var/lib/tor/services/sourcev3/hostname"}

```

kushaldas commented:
Managed to move pass these issues, and working on the branch.

redshiftzero commented:
Regarding the hostname not being generated issue you hit - since the `hostname` file won't be generated unless the tor process has been restarted, one cause could be that the handler on `Copy torrc config file.` wasn't registered (I could imagine this happens if the play ended prematurely and then was re-run). In which case, it seems wise to add the notify action to the last tasks (`Create the client auth file *`) in order to ensure that the handler gets registered. Since registering the handler multiple times won't restart tor multiple times, this seems like a more robust approach.

kushaldas commented:
Just right now:

```
 ___________________________________________________________
/ TASK [restrict-direct-access : Write Tor v3 onion service \
\ hostname files to Admin Workstation.]                     /
 -----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

failed: [app -> localhost] (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-16 15:44:47.487858', '_ansible_no_log': False, u'stdout': u'fahxp2rdvuf6mtk5hn564wx5msljnxcsdxtchfsvrctc663djan76uyd.onion', u'cmd': [u'cat', u'/var/lib/tor/services/sourcev3/hostname'], u'rc': 0, 'item': {u'service': u'sourcev3', u'filename': u'app-sourcev3-ths'}, u'delta': u'0:00:00.003192', '_ansible_item_label': {u'service': u'sourcev3', u'filename': u'app-sourcev3-ths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/sourcev3/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'fahxp2rdvuf6mtk5hn564wx5msljnxcsdxtchfsvrctc663djan76uyd.onion'], u'start': u'2019-08-16 15:44:47.484666', '_ansible_ignore_errors': None, 'failed': False}) => {"changed": false, "item": {"changed": false, "cmd": ["cat", "/var/lib/tor/services/sourcev3/hostname"], "delta": "0:00:00.003192", "end": "2019-08-16 15:44:47.487858", "failed": false, "invocation": {"module_args": {"_raw_params": "cat /var/lib/tor/services/sourcev3/hostname", "_uses_shell": false, "argv": null, "chdir": null, "creates": null, "executable": null, "removes": null, "stdin": null, "warn": true}}, "item": {"filename": "app-sourcev3-ths", "service": "sourcev3"}, "rc": 0, "start": "2019-08-16 15:44:47.484666", "stderr": "", "stderr_lines": [], "stdout": "fahxp2rdvuf6mtk5hn564wx5msljnxcsdxtchfsvrctc663djan76uyd.onion", "stdout_lines": ["fahxp2rdvuf6mtk5hn564wx5msljnxcsdxtchfsvrctc663djan76uyd.onion"]}, "msg": "AnsibleError: template error while templating string: Unexpected end of template. Jinja was looking for the following tags: 'endif'. The innermost block that needs to be closed is 'if'.. String: {% if item.item.filename.endswith('-aths') %}{{ item.stdout|truncate(56, True) }}:descriptor:x25519:{{ (lookup('file', role_path+'../../tor_v3_keys.json')|from_json)[tor_v3_service_map[item.item.filename]] }}{% else %}{{ item.stdout }}\n"}
failed: [app -> localhost] (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-16 15:44:47.607153', '_ansible_no_log': False, u'stdout': u'wzrjny42nwb66aycqehhnrge23a2s3ccdnzduhdqjdcpxncnyouvpyyd.onion', u'cmd': [u'cat', u'/var/lib/tor/services/journalistv3/hostname'], u'rc': 0, 'item': {u'service': u'journalistv3', u'filename': u'app-journalistv3-aths'}, u'delta': u'0:00:00.003228', '_ansible_item_label': {u'service': u'journalistv3', u'filename': u'app-journalistv3-aths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/journalistv3/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'wzrjny42nwb66aycqehhnrge23a2s3ccdnzduhdqjdcpxncnyouvpyyd.onion'], u'start': u'2019-08-16 15:44:47.603925', '_ansible_ignore_errors': None, 'failed': False}) => {"changed": false, "item": {"changed": false, "cmd": ["cat", "/var/lib/tor/services/journalistv3/hostname"], "delta": "0:00:00.003228", "end": "2019-08-16 15:44:47.607153", "failed": false, "invocation": {"module_args": {"_raw_params": "cat /var/lib/tor/services/journalistv3/hostname", "_uses_shell": false, "argv": null, "chdir": null, "creates": null, "executable": null, "removes": null, "stdin": null, "warn": true}}, "item": {"filename": "app-journalistv3-aths", "service": "journalistv3"}, "rc": 0, "start": "2019-08-16 15:44:47.603925", "stderr": "", "stderr_lines": [], "stdout": "wzrjny42nwb66aycqehhnrge23a2s3ccdnzduhdqjdcpxncnyouvpyyd.onion", "stdout_lines": ["wzrjny42nwb66aycqehhnrge23a2s3ccdnzduhdqjdcpxncnyouvpyyd.onion"]}, "msg": "AnsibleError: template error while templating string: Unexpected end of template. Jinja was looking for the following tags: 'endif'. The innermost block that needs to be closed is 'if'.. String: {% if item.item.filename.endswith('-aths') %}{{ item.stdout|truncate(56, True) }}:descriptor:x25519:{{ (lookup('file', role_path+'../../tor_v3_keys.json')|from_json)[tor_v3_service_map[item.item.filename]] }}{% else %}{{ item.stdout }}\n"}

```

kushaldas commented:
Fresh v2 install, and then tried to enable v3.

```
 ________________________________________________________
/ TASK [restrict-direct-access : Write Tor Onion Service \
\ hostname files to Admin Workstation.]                  /
 --------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app -> localhost] => (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-19 12:58:45.112698', '_ansible_no_log': False, u'stdout': u'xs52ooqugfjjdusv.onion', u'cmd': [u'cat', u'/var/lib/tor/services/source/hostname'], u'rc': 0, 'item': {u'service': u'source', u'filename': u'app-source-ths'}, u'delta': u'0:00:00.004463', '_ansible_item_label': {u'service': u'source', u'filename': u'app-source-ths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/source/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'xs52ooqugfjjdusv.onion'], u'start': u'2019-08-19 12:58:45.108235', '_ansible_ignore_errors': None, 'failed': False})
ok: [app -> localhost] => (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-19 12:58:45.261831', '_ansible_no_log': False, u'stdout': u'vtcigbugradhbkdg.onion cXHITfMLb12/iRI3mJ8I8R # client: journalist', u'cmd': [u'cat', u'/var/lib/tor/services/journalist/hostname'], u'rc': 0, 'item': {u'service': u'journalist', u'filename': u'app-journalist-aths'}, u'delta': u'0:00:00.004911', '_ansible_item_label': {u'service': u'journalist', u'filename': u'app-journalist-aths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/journalist/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'vtcigbugradhbkdg.onion cXHITfMLb12/iRI3mJ8I8R # client: journalist'], u'start': u'2019-08-19 12:58:45.256920', '_ansible_ignore_errors': None, 'failed': False})
 __________________________________________________________
/ TASK [restrict-direct-access : Wait for all Tor v3 onion \
\ services hostname files.]                                /
 ----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

failed: [app] (item={u'service': u'sourcev3', u'filename': u'app-sourcev3-ths'}) => {"changed": false, "elapsed": 300, "item": {"filename": "app-sourcev3-ths", "service": "sourcev3"}, "msg": "Timeout when waiting for file /var/lib/tor/services/sourcev3/hostname"}
failed: [app] (item={u'service': u'journalistv3', u'filename': u'app-journalist.auth_private'}) => {"changed": false, "elapsed": 300, "item": {"filename": "app-journalist.auth_private", "service": "journalistv3"}, "msg": "Timeout when waiting for file /var/lib/tor/services/journalistv3/hostname"}

```

kushaldas commented:
Uh, even though the site-specific has the following:

```
v2_onion_services: true
v3_onion_services: true
```
The `/etc/tor/torrc` file only has v2 details.

kushaldas commented:
It is not trying to do the install first.


```
 $ ./securedrop-admin install
INFO: Now installing SecureDrop on remote servers.
INFO: You will be prompted for the sudo password on the servers.
INFO: The sudo password is only necessary during initial installation.
SUDO password: 
 _________________________________________
< PLAY [Prepare servers for installation] >
 -----------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 _____________________________________________________________
/ TASK [prepare-servers : Ensure apt has been updated without \
\ following redirects]                                        /
 -------------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 [WARNING]: raw module does not support the environment keyword

 [WARNING]: raw module does not support the environment keyword

ok: [mon]
ok: [app]
 _________________________________________________________
/ TASK [prepare-servers : Inform user that apt traffic is \
\ being redirected]                                       /
 ---------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app] => {
    "changed": false, 
    "msg": "All assertions passed"
}
ok: [mon] => {
    "changed": false, 
    "msg": "All assertions passed"
}
 _____________________________________________________
/ TASK [prepare-servers : Install python and packages \
\ required by installer]                              /
 -----------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 [WARNING]: raw module does not support the environment keyword

 [WARNING]: raw module does not support the environment keyword

ok: [app]
ok: [mon]
 __________________________________________________________
< PLAY [Add FPF apt repository and install base packages.] >
 ----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 ________________________
< TASK [Gathering Facts] >
 ------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [mon]
ok: [app]
 ______________________________________________
< TASK [Check if install has been done before] >
 ----------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app -> None]
ok: [mon -> None]
 ______________________________________________
< TASK [include_role : restrict-direct-access] >
 ----------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 _______________________________________________________
/ TASK [restrict-direct-access : Wait for all Tor Onion \
\ Services hostname files.]                             /
 -------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app] => (item={u'service': u'source', u'filename': u'app-source-ths'})
ok: [app] => (item={u'service': u'journalist', u'filename': u'app-journalist-aths'})
 __________________________________________________________
/ TASK [restrict-direct-access : Collect Tor Onion Service \
\ hostnames.]                                              /
 ----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app] => (item={u'service': u'source', u'filename': u'app-source-ths'})
ok: [app] => (item={u'service': u'journalist', u'filename': u'app-journalist-aths'})
 ________________________________________________________
/ TASK [restrict-direct-access : Write Tor Onion Service \
\ hostname files to Admin Workstation.]                  /
 --------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

ok: [app -> localhost] => (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-19 13:20:30.752479', '_ansible_no_log': False, u'stdout': u'xs52ooqugfjjdusv.onion', u'cmd': [u'cat', u'/var/lib/tor/services/source/hostname'], u'rc': 0, 'item': {u'service': u'source', u'filename': u'app-source-ths'}, u'delta': u'0:00:00.003525', '_ansible_item_label': {u'service': u'source', u'filename': u'app-source-ths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/source/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'xs52ooqugfjjdusv.onion'], u'start': u'2019-08-19 13:20:30.748954', '_ansible_ignore_errors': None, 'failed': False})
ok: [app -> localhost] => (item={'_ansible_parsed': True, 'stderr_lines': [], '_ansible_item_result': True, u'end': u'2019-08-19 13:20:30.887810', '_ansible_no_log': False, u'stdout': u'vtcigbugradhbkdg.onion cXHITfMLb12/iRI3mJ8I8R # client: journalist', u'cmd': [u'cat', u'/var/lib/tor/services/journalist/hostname'], u'rc': 0, 'item': {u'service': u'journalist', u'filename': u'app-journalist-aths'}, u'delta': u'0:00:00.003077', '_ansible_item_label': {u'service': u'journalist', u'filename': u'app-journalist-aths'}, u'stderr': u'', u'changed': False, u'invocation': {u'module_args': {u'warn': True, u'executable': None, u'_uses_shell': False, u'_raw_params': u'cat /var/lib/tor/services/journalist/hostname', u'removes': None, u'argv': None, u'creates': None, u'chdir': None, u'stdin': None}}, 'stdout_lines': [u'vtcigbugradhbkdg.onion cXHITfMLb12/iRI3mJ8I8R # client: journalist'], u'start': u'2019-08-19 13:20:30.884733', '_ansible_ignore_errors': None, 'failed': False})
 __________________________________________________________
/ TASK [restrict-direct-access : Wait for all Tor v3 onion \
\ services hostname files.]                                /
 ----------------------------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

failed: [app] (item={u'service': u'sourcev3', u'filename': u'app-sourcev3-ths'}) => {"changed": false, "elapsed": 300, "item": {"filename": "app-sourcev3-ths", "service": "sourcev3"}, "msg": "Timeout when waiting for file /var/lib/tor/services/sourcev3/hostname"}
failed: [app] (item={u'service': u'journalistv3', u'filename': u'app-journalist.auth_private'}) => {"changed": false, "elapsed": 300, "item": {"filename": "app-journalist.auth_private", "service": "journalistv3"}, "msg": "Timeout when waiting for file /var/lib/tor/services/journalistv3/hostname"}
 ____________________
< NO MORE HOSTS LEFT >
 --------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

 ____________________
< NO MORE HOSTS LEFT >
 --------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

	to retry, use: --limit @/home/kdas/code/securedrop/install_files/ansible-base/securedrop-prod.retry
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

app                        : ok=8    changed=0    unreachable=0    failed=1   
mon                        : ok=5    changed=0    unreachable=0    failed=0   

ERROR (run with -v for more): Command '['/home/kdas/code/securedrop/install_files/ansible-base/securedrop-prod.yml', '--ask-become-pass']' returned non-zero exit status 2

```

kushaldas commented:
## Test steps

I have `app-journalist-aths` and `app-journalist.auth_private` from an old run, I manually deleted the `tor_v3_keys.json` file.


- [ ] Mark `v3_onion_services` as `false` in the `install_files/ansible-base/group_vars/all/site-specific`
- [ ] Install fresh securedrop via `./securedrop-admin install`
- [ ] Update `install_files/ansible-base/group_vars/all/site-specific` to mark `v3_onion_services` and `v2_onion_services` as `true`.
- [ ] Run `./securedrop-admin install` again. 

The last step is failing for me as I pasted the comment above. @conorsch let me know if you need more details.

conorsch commented:
Rebased on top of latest develop (9af7898c69caa8f265ee12a122c3ea2921ceff98), and squashed a few "fix" commits. Will try to reproduce @kushaldas's reported error next.

conorsch commented:
> I have `app-journalist-aths` and `app-journalist.auth_private` from an old run, I manually deleted the `tor_v3_keys.json` file.

Thanks for the detailed testing steps, @kushaldas. The scenario you described is a misconfiguration: it means the Admin copied _some_ information from the other Admin Workstation, but not all of it. The existing failure logic if _no_ information was copied currently emits the following message:

> ERROR. The 'sdconfig' settings do not specify v3 Onion Services,
but v3 Onion Services were found on the server. If your SecureDrop
instance has multiple Administrators, contact the other Administrators
to request the Tor v3 Onion Service config information. You must copy
the 'tor_v3_keys.json' and '*.auth_private' files to this workstation,
then re-run the install action.

That said, is the scenario described above one that we want to support? At some point, we're going to have to trust Admins to follow the docs—we cannot accommodate for all possible misconfigurations. If we do decide to support this scenario, then a task in the "validate" role (which runs on Tails) seems the best approach to detect that v3-client-auth files exist, but v3-keypairs do not. Still, that feels like quite an edge case to me. cc @redshiftzero for input.

redshiftzero commented:
I think it'd be a nice to have to add a task in the validate role to fail with an understandable error message if `tor_v3_keys.json` is missing and the v3 ths files are present - but let's do in a followup ticket so we can get the required basic functionality in for the case where the admin follows the docs

conorsch commented:
Pushed up one more rebase to include #4680. 

> but let's do in a followup ticket

Filed, see #4681.

kushaldas commented:
Still fails for prod vms, same issue with `restrict-direct-access` and Tor v3 service hostname files.

conorsch commented:
Latest round of changes provides coverage for the ssh-over-lan config case. Satisfied with functionality here. We now have conflicts on the requirements, however, so will try to sort that out before submitting approval.

conorsch commented:
Rebased on top of latest develop (a9087a2d42f2cb6a645adbb6d09f204739c4a5eb), to accommodate for changes from #4665. Specifically, these were minor discrepancies in the pip deps:

* `setuptools`, `41.1.0` -> `41.2.0`
* `requests`, `2.20.0` -> `2.22.0`

Now that conflicts are resolved, let's make sure CI passes, then we're good to go.
